File: src/msg_sub.cpp
#include <ros/ros.h>
#include "bingda_prictices/student.h"

void studentInfoCallback(const bingda_prictices::student::ConstPtr& msg) // 定义一个回调函数，参数是一个指向我们定义的消息类型的常量指针
{
    ROS_INFO("Received student info: %s %s, age: %d, score: %d", msg->firstname.c_str(), msg->lastname.c_str(), msg->age, msg->score); // 打印接收到的消息内容，使用ROS_INFO宏来输出日志信息
}

int main(int argc, char **argv) // ROS节点的入口函数
{
    ros::init(argc, argv, "msg_sub"); // 初始化ROS节点，节点名称为msg_sub
    ros::NodeHandle n; // 创建一个NodeHandle对象，负责与ROS系统进行通信
    ros::Subscriber sub = n.subscribe("student_info", 1000, studentInfoCallback); // 创建一个Subscriber对象，订阅名为"student_info"的主题，队列大小为1000，回调函数是studentInfoCallback

    ros::spin(); // 进入循环，等待回调函数被调用

    return 0;
}

File: scripts/msg_sub.py
#!/usr/bin/env python3
import rospy
from std_msgs.msg import String
from bingda_prictices.msg import student  # 导入我们自己定义的消息类型



def msg_sub():  # 定义一个函数，名字叫msg_sub，用来订阅消息并处理接收到的消息
    rospy.init_node('msg_sub', anonymous=True)  # 初始化ROS节点，节点名称为'msg_sub'，anonymous=True表示如果有同名节点会自动添加随机数后缀以区分
    rospy.Subscriber("student_info", student, callback)  # 创建一个订阅者对象，订阅名为'student_info'的话题，消息类型为student，回调函数为callback
    rospy.spin()  # 保持程序持续运行，等待回调函数被调用


def callback(data): # 定义一个回调函数，名字叫callback，当接收到消息时会调用这个函数，参数data就是接收到的消息对象
    rospy.loginfo("Received student info: %s %s, age: %d, score: %d", 
                  data.firstname, data.lastname, data.age, data.score)  # 在ROS日志中输出接收到的消息内容
    


if __name__ == '__main__':
    msg_sub()  # 调用msg_sub函数开始订阅消息
